### DISCUSSION TITLE
Vision of working with GitHub actions.

### DISCUSSION THREAD

**Discussion by andrii-kl** [icon-project/icon-bridge#137] [Ideas]
### Examples
**Our current approach:** 
https://github.com/icon-project/icon-bridge/actions
<img width="301" alt="Screen Shot 2022-07-28 at 12 31 36 PM" src="https://user-images.githubusercontent.com/18900364/181473191-706bf9c4-7c2f-4ae3-ae99-5013deb1ce7e.png">
<img width="304" alt="Screen Shot 2022-07-28 at 12 31 45 PM" src="https://user-images.githubusercontent.com/18900364/181473207-f651eaad-889b-493e-b803-b7d4805f1f64.png">

**Other projects:** 
https://github.com/eslint/eslint/actions/workflows/ci.yml

<img width="298" alt="Screen Shot 2022-07-28 at 1 04 38 PM" src="https://user-images.githubusercontent.com/18900364/181479729-7fac877f-f852-41b2-9e71-288cff3b8f15.png">

https://github.com/axios/axios/actions

<img width="310" alt="Screen Shot 2022-07-28 at 1 05 13 PM" src="https://user-images.githubusercontent.com/18900364/181479841-1f8ab4ba-3d36-44e3-b07d-923f0710fe6e.png">

https://github.com/strapi/strapi/actions

### My vision of icon bridge GitHub actions workflows structure
- Build
- Release
- Integration tests

### Workflows description

1. Build
 It should be lightweight workflow.  We don't use docker images or other time consuming process.
 [build.yml](https://github.com/icon-project/icon-bridge/blob/main/.github/workflows/build.yml)

Jobs: 
    a) build_bsc  (solidity)
    b) build_icon  (javascore)
    c) build_bmr    (go)

 **Execution process:** Run for each PR and for each merge to the main or develop branch.
 
 **Functionality:** build | compile | run unit tests
 

 2. Release 
   
 In progress...
 
 
 3. Integration tests
 It's a time consuming workflow. We can use docker images to simplify configuration process.
 
 **Functionality:** run integration tests

**Comment by MuhammedIrfan** [comment]
Further Reference

https://buildkite.com
https://buildkite.com/nearprotocol/nearcore

**Comment by CyrusVorwald** [comment]
Here are some comparable repos in the crypto space worth looking at:

- https://github.com/certusone/wormhole
- https://github.com/nomad-xyz/monorepo
- https://github.com/nomad-xyz/rust
- https://github.com/cosmos/ibc-go
- https://github.com/paritytech/parity-bridges-common
- https://github.com/LayerZero-Labs/LayerZero
- https://github.com/aurora-is-near/rainbow-bridge

**Comment by CyrusVorwald** [comment]
It is actually common for there to be many CI workflows in a monorepo. For example see https://github.com/npm/cli/tree/latest/.github/workflows
